keps/sig-apps/961-maxunavailable-for-statefulset/README.md

```diff
@@ -106,152 +306,194 @@ type RollingUpdateStatefulSetStrategy struct {
+    // (MaxUnavailable - Unavailable) Pods, in order with respect to their ordinal for termination. Delete
+    // those pods and count the successful deletions. Update the status with the correct number of deletions.
+    unavailablePods := 0
+    for target := len(replicas) - 1; target >= 0; target-- {
+    //////////////////// 
+    ////////////////////
+    // New Check here //
+    //////////////////// 
+    ////////////////////
+    if !isRunningAndAvailable(replicas[target], set.Spec.MinReadySeconds) || isTerminating(replicas[target]) {
+        unavailablePods++
+      }
+    }
+
+    if unavailablePods >= maxUnavailable {
+      logger.V(2).Info("StatefulSet found unavailablePods, more than or equal to allowed maxUnavailable",
+        "statefulSet", klog.KObj(set),
+        "unavailablePods", unavailablePods,
+        "maxUnavailable", maxUnavailable)
+      return &status, nil
+    }
+
+    // Now we need to delete MaxUnavailable- unavailablePods
+    // start deleting one by one starting from the highest ordinal first
+    podsToDelete := maxUnavailable - unavailablePods
+
+    deletedPods := 0
+    for target := len(replicas) - 1; target >= updateMin && deletedPods < podsToDelete; target-- {
+
+      // delete the Pod if it is healthy and the revision doesnt match the target
+      if getPodRevision(replicas[target]) != updateRevision.Name && !isTerminating(replicas[target]) {
+        // delete the Pod if it is healthy and the revision doesnt match the target
+        logger.V(2).Info("StatefulSet terminating Pod for update",
+          "statefulSet", klog.KObj(set),
+          "pod", klog.KObj(replicas[target]))
+        if err := ssc.podControl.DeleteStatefulPod(set, replicas[target]); err != nil {
+          if !errors.IsNotFound(err) {
+            return &status, err
+          }
+        }
+        deletedPods++
+        status.CurrentReplicas--
+      }
+    }
+    return &status, nil
 
 	}
 ...
 ```
 
-### Risks and Mitigations
-We are proposing a new field called `maxUnavailable` whose default value will be 1. In this mode, StatefulSet will behave exactly like its current behavior.
-Its possible we introduce a bug in the implementation. The mitigation currently is that is disabled by default in Alpha phase for people to try out and give
-feedback. 
-In Beta phase when its enabled by default, people will only see issues or bugs when `maxUnavailable` is set to something greater than 1. Since people have
-tried this feature in Alpha, we would have time to fix issues.
+#### Metrics
+
+Similar to what we have for deployments, we can add a new metric named `kube_statefulset_spec_strategy_rollingupdate_max_unavailable` which just lets users know the maximum number of unavailable replicas during a rolling update of a StatefulSet. This would be add to kube-state-metrics.
+
+<!-- need to decide if we keep and can use statefulset_unavailability_violation -->
```

Comment:
Just adding comments here in the PR for me to know what to update if we come to a decision here.